### PR TITLE
#41: Improve handling of table view row heights

### PR DESCRIPTION
**Pull request by asmallteapot**
## Changes in this pull request

Issue fixed: #41

### Checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [x] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [x] I have reviewed the [contributing guide](https://github.com/plangrid/ReactiveLists/blob/master/.github/CONTRIBUTING.md)

### Test failure

Okay, I lied :innocent: `TableViewDriverTests.testTableViewRows()` is failing for the `42` case and it's not at all clear why.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,15 @@ The changelog for `ReactiveLists`. Also see the [releases](https://github.com/pl
 
 ------
 
+0.2.0
+-----
+
+- Improve semantics of row height API
```
Comment (jessesquires) [inline CHANGELOG.md]: let's link the issue number here and author:

- Improve semantics of row height API ([#44](https://github.com/plangrid/ReactiveLists/issues/41), [@asmallteapot](https://github.com/asmallteapot))

```markdown
- Improve semantics of row height API ([#44](https://github.com/plangrid/ReactiveLists/issues/41), [@asmallteapot](https://github.com/asmallteapot))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,15 @@ The changelog for `ReactiveLists`. Also see the [releases](https://github.com/pl
...
+	- Make `TableCellViewModel.rowHeight` optional, defaulting to `nil`
+	- Add `TableViewModel.defaultRowHeight`, defaulting to `44.0`
+
+------
```
Comment (jessesquires) [inline CHANGELOG.md]: can remove this `-----`

File: Sources/TableViewDriver.swift [hunk 3]
```diff
@@ -255,7 +255,7 @@ extension TableViewDriver: UITableViewDelegate {
 
     /// :nodoc:
     public func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
-        return self.tableViewModel?[indexPath]?.rowHeight ?? 44
+        return self.tableViewModel?[indexPath]?.rowHeight ?? self.tableViewModel?.defaultRowHeight ?? 44
```
Comment (jessesquires) [inline Sources/TableViewDriver.swift]: By the time this method is called, we should have a non-nil `tableViewModel` (*I think*)

can we adjust to:

```swift
guard let tableViewModel = self.tableViewModel else { return 0 }
return tableViewModel[indexPath]?.rowHeight ?? tableViewModel.defaultRowHeight
```

(eventually, i think we'll make `self.tableViewModel` non-optional)

File: Sources/TableViewModel.swift [hunk 4]
```diff
@@ -194,6 +194,9 @@ public struct TableSectionViewModel {
 /// The view model that describes a `UITableView`.
 public struct TableViewModel {
 
+    /// The default row height for this table view.
```
Comment (jessesquires) [inline Sources/TableViewModel.swift]: > /// The default row height for this table view. Default value is `44`.

**Comment by jessesquires** [comment]
Hey @asmallteapot 👋 

Are you still interested in working on this? `master` is all updated now and green ✅ 

If you're too busy, I'll carry this forward 😊 (and will make sure you get the commit credits 😉)

**Comment by asmallteapot** [comment]
Just pushed a rebase; planning to update with your feedback (and any needed test fixes) this weekend. Thanks for your help!

**Comment by PlanBot-iOS** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">All pull requests should have a milestone attached</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by anayini** [comment]
@asmallteapot - Can I update this for you?

**Comment by asmallteapot** [comment]
@anayini yeah, go for it

**Comment by jessesquires** [comment]
Thanks again @asmallteapot !

@anayini carried this forward in #154. we preserved your commits and you've been credited in the changelog 😄

**Pull request closed without merging**